spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -112,6 +112,246 @@ Subsequently, an implementation of Jakarta Data, specifically tailored to the ch
+For example, a repository interface may inherit the `CrudRepository` interface defined by this specification.
+
+[NOTE]
+A Jakarta Data provider might go beyond what is required by this specification and support abstract methods which do not fall into any of the above categories.
+Such functionality is not defined by this specification, and so programs with repositories which declare such methods are not portable between providers.
+
+The subsections below specify the rules that an abstract method declaration must observe so that the Jakarta Data implementation is able to provide an implementation of the abstract method.
+
+- If every abstract method of a repository complies with the rules specified below, then the Jakarta Data implementation must provide an implementation of the repository.
+- Otherwise, if a repository declares an abstract method which does not comply with the rules specified below, or makes use of functionality which is not supported by the Jakarta Data implementation, then an error might be produced by the Jakarta Data implementation at build time or at runtime.
+
+The portability of a given repository interface between Jakarta Data implementations depends on the portability of the entity types it uses.
+If an entity class is not portable between given implementations, then any repository which uses the entity class is also unportable between those implementations.
+
+[NOTE]
+Additional portability guarantees may be provided by specifications which extend this specification, specializing to a given class of datastore.
+
+==== Lifecycle methods
+
+A _lifecycle method_ is an abstract method annotated with a _lifecycle annotation_.
+Lifecycle methods allow the program to make changes to persistent data in the data store.
+
+A lifecycle method must:
+
+- have a single parameter annotated with a lifecycle annotation, whose type is a concrete entity class `E`, and
+- be declared `void`, or return exactly the same entity type `E`.
+
+That is, its signature must be:
+
+[source,java]
+----
+void lifecycle(@Lifecycle Entity e);
+----
+
+or:
+
+[source,java]
+----
+Entity lifecycle(@Lifecycle Entity e);
+----
+
+where `lifecycle` is the arbitrary name of the method, `Entity` is a concrete entity class, and `Lifecycle` is a lifecycle annotation.
+
+This specification defines three built-in lifecycle annotations: `@Insert`, `@Update`, and `@Delete`.
+
+For example:
+
+[source,java]
+----
+void insertBook(@Insert Book book);
+----
+
+Lifecycle methods are not guaranteed to be portable between all providers.
+
+[NOTE]
+Any Jakarta Data implementation is strongly encouraged to support lifecycle methods declared using the built-in lifecycle annotations, to enable portability of user code between Jakarta Data providers.
+However, since these lifecycle operations are not universal across all possible data stores, support for `@Insert`, `@Update`, and `@Delete` is not required by this specification.
```

Comment:
```suggestion
Jakarta Data providers must support lifecycle methods to the extent that the data store is capable of the corresponding operation. If the data store is not capable of the operation, the Jakarta Data provider must raise `UnsupportedOperationException` when the operation is attempted, per the requirements of the JavaDoc for the lifecycle annotation.
```

Reply:
@njr-11 I'm fine with this change, but I would like it to also mention that a provider might alternatively choose to detect and report the error at compile time.

Reply:
> @njr-11 I'm fine with this change, but I would like it to also mention that a provider might alternatively choose to detect and report the error at compile time.

Good point. I added that to the code suggestion.